appmonitoring/ts/src/CertificateGenerator.ts

```diff
@@ -23,113 +19,125 @@ export class CertificateManager {
-            caCert.publicKey = keys.publicKey;
-            caCert.privateKey = keys.privateKey;
-            const timeNowNum: number = Date.now()
-            caCert.validity.notBefore = new Date(timeNowNum - (5 * 60 * 1000)); //5 Mins ago
-            caCert.validity.notAfter = new Date(timeNowNum + (2 * 365 * 24 * 60 * 60 * 1000)); //2 Years from now
-
-            const attributes = [{
-                shortName: 'CN',
-                value: 'applicationinsights-ca'
-            }];
-            caCert.setSubject(attributes);
-            caCert.setIssuer(attributes);
-        
-            const extensions = [{
-                name: 'basicConstraints',
-                cA: true
-            },
-            {
-                name: 'subjectKeyIdentifier',
-                keyIdentifier: caCert.generateSubjectKeyIdentifier().getBytes(),
-            },
-            {
-                name: 'keyUsage',
-                keyCertSign: true,
-                cRLSign: true,
-                digitalSignature: true,
-                keyEncipherment: true,
-            }];
-
-            caCert.setExtensions(extensions);
-            caCert.sign(caCert.privateKey,forge.md.sha256.create());
-
-            const caCertResult: CACertData = {
-                certificate: forge.pki.certificateToPem(caCert),
-                serviceKey: forge.pki.privateKeyToPem(caCert.privateKey)
-            }
+    private static async GenerateCACertificate(existingKeyPair?: forge.pki.rsa.KeyPair): Promise<forge.pki.Certificate> {
+        const currentTime: number = Date.now();
+        const caCert = forge.pki.createCertificate();
+        const keys = existingKeyPair || forge.pki.rsa.generateKeyPair();
+        caCert.serialNumber = CertificateManager.randomHexSerialNumber();
+        caCert.publicKey = keys.publicKey;
+        caCert.privateKey = keys.privateKey;
+        caCert.validity.notBefore = new Date(currentTime - (5 * 60 * 1000)); //5 Mins ago
+        caCert.validity.notAfter = new Date(currentTime + (2 * 365 * 24 * 60 * 60 * 1000)); //2 Years from now
 
-            const host_attributes = [{
-                shortName: 'CN',
-                value: WebhookDNSEndpoint
-            }];
-        
-            const host_extensions = [{
-                name: 'basicConstraints',
-                cA: false
-            }, 
-            {
-                name: 'authorityKeyIdentifier',
-                keyIdentifier: caCert.generateSubjectKeyIdentifier().getBytes(),
-            }, 
-            {
-                name: 'keyUsage',
-                digitalSignature: true,
-                keyEncipherment: true
-            },
-            {
-                name: 'extKeyUsage',
-                serverAuth: true
-            }, 
-            {
-                name: 'subjectAltName',
-                altNames: [{ type: 2, value: WebhookDNSEndpoint }]
-            }];
+        const attributes = [{
```

Comment:
This requires a lot of unit tests. I've made unit testing somewhat streamlined, so add to the file where tests for CertificateGenerator live please. Going forward we'll have unit tests for everything we do, otherwise we'll be in trouble big time :)